password_enhancements.module

```diff
@@ -0,0 +1,471 @@
+  /** @var \Drupal\user\UserInterface $user */
+  if (!empty($form['account']['pass'])) {
+    /** @var \Drupal\Core\Entity\EntityFormInterface $form_object */
+    $form_object = $form_state->getFormObject();
+    $user = $form_object->getEntity();
+
+    try {
+      $user->setPassword($form_state->getUserInput()['pass']['pass1'])->save();
+    }
+    catch (Exception $e) {
+      Drupal::service('logger.password_enhancements')->exception($e);
+      Drupal::messenger()->addError(t('Failed to set your password, please <a href="@url">reset your password</a>.', [
+        '@url' => Url::fromRoute('user.pass')->toString(),
+      ]));
+    }
+  }
+}
+
+/**
+ * Ajax callback for updating the constraints on role change.
+ *
+ * @param array $form
+ *   The form.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   The current form state.
+ *
+ * @return array
+ *   Renderable form array for password policy constraints.
+ */
+function password_enhancements_update_constraints_by_roles(array $form, FormStateInterface $form_state): array {
+  return $form['account']['password_policy_constraints'];
+}
+
+/**
+ * Implements hook_entity_base_field_info().
+ */
+function password_enhancements_entity_base_field_info(EntityTypeInterface $entity_type): array {
+  $fields = [];
+
+  if ($entity_type->id() === 'user') {
+    $fields['password_enhancements_password_changed_date'] = BaseFieldDefinition::create('datetime')
+      ->setLabel(t('Last password change'))
+      ->setDescription(t('The last time when the user has changed the password.'))
+      ->setDefaultValue(0);
+
+    $fields['password_enhancements_password_change_required'] = BaseFieldDefinition::create('boolean')
+      ->setLabel(t('Password change required'))
+      ->setDescription(t('Forces the user to change the password.<br><strong>Warning:</strong> If it is checked, the user will be redirected to the password change form.'))
+      ->setDefaultValue(0)
+      ->setDisplayOptions('form', [
+        'type' => 'boolean_checkbox',
+        'settings' => [
+          'display_label' => TRUE,
+        ],
+        'weight' => 99,
+      ]);
+  }
+
+  return $fields;
+}
+
+/**
+ * Implements hook_ENTITY_TYPE_presave().
+ */
+function password_enhancements_user_presave(UserInterface $user): void {
+  // Update the last password changed date if the password was changed.
+  if (empty($user->original) || $user->original->pass->value !== $user->pass->value) {
+    $user->password_enhancements_password_changed_date->setValue(Drupal::time()->getRequestTime());
+  }
+}
+
+/**
+ * Implements hook_user_login().
+ */
+function password_enhancements_user_login(UserInterface $user): void {
+  $policy = Drupal::service('entity_type.manager')->getStorage('password_enhancements_policy')->loadByRoleAndPriority($user->getRoles());
+  if ($user->password_enhancements_password_change_required->value || ($policy !== NULL && Drupal::service('password_enhancements.password_checker')->isExpired($policy))) {
```

Comment:
The same nit as above.
![image](https://user-images.githubusercontent.com/1755573/60488514-76022d80-9ca2-11e9-8ee0-516c767c0153.png)

Reply:
I would use the same, although getting the value like this is easier than calling `$user->get('password_enhancements_password_changed_date')->getValue()` because this latter returns with an array while the former returns the value directly. I know it's ugly but works.